src/pd/client.rs

```diff
@@ -11,247 +11,408 @@
+    where F: Fn(&pdpb_grpc::PDClient) -> result::Result<R, grpc::error::GrpcError>
+{
+    let mut resp = None;
+    for _ in 0..MAX_RETRY_COUNT {
+        let r = {
+            let inner = client.inner.read().unwrap();
+            let timer = PD_SEND_MSG_HISTOGRAM.start_timer();
+            let r = f(&inner.client);
+            timer.observe_duration();
+            r
+        };
+
+        match r {
+            Ok(r) => {
+                resp = Some(r);
+                break;
+            }
+            Err(e) => {
+                error!("fail to request: {:?}", e);
+                let mut inner = client.inner.write().unwrap();
+                match try_connect_leader(&inner.members) {
+                    Ok((cli, mbrs)) => {
+                        inner.client = cli;
+                        inner.members = mbrs;
+                    }
+                    Err(e) => {
+                        error!("fail to connect to PD leader {:?}", e);
+                        thread::sleep(Duration::from_secs(RETRY_INTERVAL));
+                    }
+                }
                 continue;
             }
+        }
+    }
 
-            let mut stream = self.stream.take().unwrap();
-            // We may send message to a not leader pd, retry.
+    resp.ok_or(box_err!("fail to request"))
+}
 
-            let (id, resp) = match send_msg(&mut stream, msg_id, req) {
-                Err(e) => {
-                    warn!("send message to pd failed {:?}", e);
-                    // TODO: figure out a better way to do backoff
-                    thread::sleep(Duration::from_millis(50));
-                    continue;
-                }
-                Ok((id, resp)) => (id, resp),
-            };
+fn check_resp_header(header: &pdpb::ResponseHeader) -> Result<()> {
+    if !header.has_error() {
+        return Ok(());
+    }
+    // TODO: translate more error types
+    let err = header.get_error();
+    Err(box_err!(err.get_message()))
+}
 
-            if id != msg_id {
-                return Err(box_err!("pd response msg_id not match, want {}, got {}", msg_id, id));
-            }
+impl fmt::Debug for RpcClient {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt,
+               "PD gRPC Client connects to cluster {:?}",
+               self.cluster_id)
+    }
+}
 
-            self.stream = Some(stream);
+impl PdClient for RpcClient {
+    fn get_cluster_id(&self) -> Result<u64> {
+        Ok(self.cluster_id)
+    }
 
-            return Ok(resp);
-        }
+    fn bootstrap_cluster(&self, stores: metapb::Store, region: metapb::Region) -> Result<()> {
+        let mut req = pdpb::BootstrapRequest::new();
+        req.set_header(self.header());
+        req.set_store(stores);
+        req.set_region(region);
 
-        Err(box_err!("send message to pd failed"))
+        let resp = try!(do_request(self, |client| client.Bootstrap(req.clone())));
+        try!(check_resp_header(resp.get_header()));
```

Comment:
I have also considered, let's look at 
`fn do_request<F, R>(client: &RpcClient, f: F) -> Result<R>`,
`R` here is the return type (a Response), if we want check headers in `do_request`, those response must `impl` some traits that have a method `get_header`, unfortunately here is no such trait.

BTW, a workaround would be macro, it will do the magic.